Async load gridicons

flootr opened the pull request:
• 61eae9a creates a separate chunk for each gridicon
• f582517 creates a separate `vendor-gridicons` chunk which gathers imported gridicons
• bef8e90 creates a separate `gridicons~build` chunk which includes the whole gridicon module

flootr committed:
try: async load gridicon chunks

flootr committed:
use a descriptive name for gridicon chunks

flootr committed:
try: webpackMode 'lazy-once'

flootr committed:
try: separate gridicon chunk (via cacheGroups)

flootr committed:
Revert "try: separate gridicon chunk (via cacheGroups)"

This reverts commit bef8e90525abb22725d685c2561a31f75c50f0b0.

flootr committed:
Revert "try: webpackMode 'lazy-once'"

This reverts commit f582517793edfd1d46410e73ae76a5d90a6079bf.

dmsnell commented:
Since it came up, how would this deal with modules that don't exist, such as when we try to load a slug that isn't in the library?

Further, this ends up loading our SVGs as JavaScript files and I wonder if that's a negative implication? We can't see the SVGs load in the developer tools as images anymore and we can't preview them. It appears to send the object structure so even the SVG text is lost in here.

flootr commented:
> Since it came up, how would this deal with modules that don't exist, such as when we try to load a slug that isn't in the library?

We probably need to provide a fallback. In case an icon doesn't exist Webpack should throw an error which we should be able to `.catch()`.

> Further, this ends up loading our SVGs as JavaScript files and I wonder if that's a negative implication? 

This is no different than the current solution, isn't it? I agree that loading SVGs directly would be a preferable solution. In the end this PR is only an experiment.

flootr committed:
AsyncGridicon: pass on props & remove placeholder

flootr committed:
AsyncGridicon: provide 'empty' fallback icon

flootr committed:
Revert "AsyncGridicon: provide 'empty' fallback icon"

This reverts commit 2fba28335f02fee1e058db7ee1b04fb3704a205e.

flootr committed:
Revert "Revert "AsyncGridicon: provide 'empty' fallback icon""

This reverts commit 4f108f7eff22587f739d7b8c329b6bf57aa58541.

flootr commented on client/components/async-gridicons/index.jsx:
```diff
@@ -9,14 +9,18 @@ import React from 'react';
+import FallbackIcon from './fallback';
 
 function AsyncGridicon( { icon = '', ...rest } ) {
 	return (
```
This is marked red, not sure why though 🤔 

<img width="186" alt="screen shot 2018-09-24 at 18 01 22" src="https://user-images.githubusercontent.com/9202899/45963643-ded9bc80-c023-11e8-8673-f9f31d7ec785.png">

flootr commented on client/components/async-gridicons/index.jsx:
```diff
@@ -0,0 +1,31 @@
+import FallbackIcon from './fallback';
+
+function AsyncGridicon( { icon = '', ...rest } ) {
+	return (
```
This is marked red, not sure why though 🤔 

<img width="186" alt="screen shot 2018-09-24 at 18 01 22" src="https://user-images.githubusercontent.com/9202899/45963643-ded9bc80-c023-11e8-8673-f9f31d7ec785.png">

blowery committed:
use the fallback icon as the placeholder to prevent the render from shifting

flootr commented on client/components/async-gridicons/index.jsx:
```diff
@@ -22,7 +22,7 @@ function AsyncGridicon( { icon = '', ...rest } ) {
 					} );
 			} }
-			placeholder={ null }
+			placeholder={ <FallbackIcon /> }
```
Ha! Good catch @blowery 😄

blowery commented:
> creates a separate vendor-gridicons chunk which gathers imported gridicons

how big was that chunk?

blowery committed:
make this a proper function so it has a name in dev tools. fix usage of class -> className

blowery committed:
a different take on the loader.
AsyncLoad isn't really meant to be passed a function. It works mostly by chance, not intent. Instead, use import directly and force a re-render when the icon is available, or fails to load.

blowery commented:
@adlk another odd desktop build failure...

blowery commented:
ICFY report http://iscalypsofastyet.com/branch?branch=try/async-gridicon

vendors~build down 20k zipped, build up 3k(?), manifest up 4.8k, net savings of 12k. 

Going to try again with all the gridicons in one chunk. Might mitigate the manifest and build increases.

blowery committed:
try all in one chunk

flootr commented:
@blowery in general, yes.

> vendors~build down 20k zipped, build up 3k(?), manifest up 4.8k, net savings of 12k.

Do you know where the 3k up comes from?

> OK, gridicons in one chunk: 2.6k increase in build, 32 byte increase in manifest, same 20K drop in vendors~build. So a 18k drop instead of 12k. I'll take that.

Do we need to take into account how often a user needs to download that gridicon chunk? Caching is probably more efficient using single chunks but gridicons also don't change often so it should be okay.

blowery commented:
No idea where the 3k was coming from.

> Do we need to take into account how often a user needs to download that gridicon chunk? Caching is probably more efficient using single chunks but gridicons also don't change often so it should be okay.

I think it'll be ok. Gridicons change very rarely.

blowery commented:
I did some extra investigation to exclude examples and index, but it seems they're naturally excluded? I'm not sure why that is, but it appears that way.

I think this is good to go.

sirreal commented:
After revert, a second attempt was landed #27527
